clang/lib/Headers/aiev2_core.h

```diff
@@ -1566,4 +1566,300 @@ inline v16accfloat addmsc_4x8_8x4( v32float v1, v32float v2, v16accfloat acc1, v
 
 #endif
 
+enum class aie_dm_resource {
+  none,
+  PM,
+  DM,
+  DM_test,
+  stack,
+  a,
+  b,
+  c,
+  d,
+  ab,
+  ac,
+  ad,
+  bc,
+  bd,
+  cd,
+  TM
+};
+
+#define __aie_dm_resource_PM                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::PM))))
+#define __aie_dm_resource_DM                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::DM))))
+#define __aie_dm_resource_test                                                 \
+  __attribute__((                                                              \
+      address_space(static_cast<unsigned>(aie_dm_resource::DM_test))))
+#define __aie_dm_resource_stack                                                \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::stack))))
+#define __aie_dm_resource_a                                                    \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::a))))
+#define __aie_dm_resource_b                                                    \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::b))))
+#define __aie_dm_resource_c                                                    \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::c))))
+#define __aie_dm_resource_d                                                    \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::d))))
+#define __aie_dm_resource_ab                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::ab))))
+#define __aie_dm_resource_ac                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::ac))))
+#define __aie_dm_resource_ad                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::ad))))
+#define __aie_dm_resource_bc                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::bc))))
+#define __aie_dm_resource_bd                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::bd))))
+#define __aie_dm_resource_cd                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::cd))))
+#define __aie_dm_resource_TM                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::TM))))
+
+template <typename T> struct aie_dm_resource_remove {
+  using type = T;
+};
+
+template <typename T, int N>
```

Comment:
Above we cast to unsigned, here we use int as address space identifier. Why the difference? (I guess the enum constants would evaluate to int, which then would not need a cast?)

Reply:
have updated it to signed, we would still need a cast for the define to be used in the AIE API or user code.